monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to parse echo timeouts")
+	}
+
+	updated, err := c.updateStats(echoRequests, echoTimeouts)
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to update nethealth stats")
+	}
+
+	return c.verifyNethealth(updated)
+}
+
+// getNethealthAddr returns the address of the local nethealth pod.
+func (c *nethealthChecker) getNethealthAddr(pods *v1.PodList) (string, error) {
+	// Find nethealth pod with matching host ip address.
+	for _, pod := range pods.Items {
+		if pod.GetLabels()[nethealthLabel] != nethealthValue {
+			continue
+		}
+		if pod.Status.HostIP == c.AdvertiseIP {
+			return fmt.Sprintf("http://%s:%d", pod.Status.PodIP, c.NethealthPort), nil
+		}
+	}
+	return "", trace.NotFound("unable to find local nethealth pod")
+}
+
+// fetchMetrics collects the network metrics from the nethealth pod.
+func fetchMetrics(ctx context.Context, addr string) (metricFamilies map[string]*dto.MetricFamily, err error) {
+	client, err := roundtrip.NewClient(addr, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to connect to nethealth service at %s.", addr)
+	}
+
+	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	var parser expfmt.TextParser
+	metricsFamilies, err := parser.TextToMetricFamilies(resp.Reader())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return metricsFamilies, nil
+}
+
+// updateStats updates netStats with new data collected in the provided
+// MetricFamily.
+func (c *nethealthChecker) updateStats(echoRequests, echoTimeouts map[string]float64) (updated []string, err error) {
+	if len(echoRequests) != len(echoTimeouts) {
+		return updated, trace.BadParameter("recieved %d timeout counters and %d request counters",
+			len(echoRequests), len(echoTimeouts))
+	}
+
+	for peer, requests := range echoRequests {
+		timeouts, ok := echoTimeouts[peer]
+		if !ok {
+			return updated, trace.BadParameter("echo timeout data not available for %s", peer)
+		}
+
+		data, err := c.peerStats.Get(peer)
+		if err != nil {
+			return updated, trace.Wrap(err)
+		}
+
+		// Calcuate counter increase since last check and replace previous total
+		requestInc := requests - data.prevRequestTotal
+		data.prevRequestTotal = requests
+		timeoutInc := timeouts - data.prevTimeoutTotal
+		data.prevTimeoutTotal = timeouts
+
+		// Request counter should be constantly increasing, if this is not the
+		// case, nethealth pod most likely reset the counter. Do no update
+		// packet loss percentages this round.
+		if requestInc <= 0 {
+			log.Warn("Request counter did not increase. Nethealth pod may have restarted.")
+			continue
+		}
+
+		// It should not be possible for the timeout counter to have increased
+		// more than the request counter. Log and ignore this situation.
+		if timeoutInc > requestInc {
+			log.WithField("request inc", requestInc).
```

Comment:
Nit: We usually make log fields w/o spaces, "request-increase", "timeout-increase".